qutip/solver/correlation.py

```diff
@@ -0,0 +1,497 @@
+        the element `0`.
+    c_ops : list
+        list of collapse operators, may be time-dependent for solver choice of
+        `me`.
+    a_op : Qobj
+        operator A.
+    args : dict
+        Dictionary of arguments to be passed to solver.
+    solver : str
+        choice of solver (`me` for master-equation and
+        `es` for exponential series).
+    options : dict
+        Options for the solver.
+
+    Returns
+    -------
+    g2, G2 : tuple
+        The normalized and unnormalized second-order coherence function.
+
+    """
+    solver = _make_solver(H, c_ops, args, options, solver)
+
+    # first calculate the photon number
+    if state0 is None:
+        state0 = steadystate(H, c_ops)
+        n = np.array([expect(state0, a_op.dag() * a_op)])
+    else:
+        n = solver.run(state0, taulist, e_ops=[a_op.dag() * a_op]).expect[0]
+
+    # calculate the correlation function G2 and normalize with n to obtain g2
+    G1 = correlation_3op(solver, state0, [0], taulist,
+                         a_op.dag(), a_op.dag() * a_op, a_op)
+
+    g2 = G2 / (n[0] * n)
+    return g2, G2
+
+
+def _fix_op_type(op, dims):
+    """ Make op QobjEvo. """
+    if op in [None, 1]:
+        op = qeye(dims)
+    return QobjEvo(op)
+
+
+def _make_solver(H, c_ops, args, options, solver):
+    H = QobjEvo(H, args=args)
+    c_ops = [QobjEvo(c_op, args=args) for c_op in c_ops]
+    if solver == "me":
+        solver_instance = MeSolver(H, c_ops, options=options)
+    elif solver == "es":
+        options = {**options, **{"method": "diag"}}
+        solver_instance = MeSolver(H, c_ops, options=options)
+    elif solver == "mc":
+        raise ValueError("MC solver for correlation has been removed")
+    return solver_instance
+
+
+def correlation_3op(solver, state0, tlist, taulist, A=None, B=None, C=None):
+    r"""
+    Calculate the three-operator two-time correlation function:
+    :math:`\left<A(t)B(t+\tau)C(t)\right>`
+
+    Note: it is not possibly to calculate a physically meaningful correlation
+    of this form where :math:`\tau<0`.
+
+    Parameters
+    ----------
+    solver : :class:`MeSolver`, :class:`BRSolver`, :class:`HEOMSolver`
+        Qutip solver for an open system.
+    state0 : Qobj
+        Initial state density matrix :math:`\rho(t_0)` or state vector
+        :math:`\psi(t_0)`.
+    tlist : array_like
+        List of times for :math:`t`. tlist must be positive and contain the
+        element `0`.
+    taulist : array_like
+        List of times for :math:`\tau`. taulist must be positive and contain
+        the element `0`.
+    A : Qobj, QobjEvo, optional, default=None
+        Operator A.
+    B : Qobj, QobjEvo, optional, default=None
+        Operator B.
+    C : Qobj, QobjEvo, optional, default=None
+        Operator C.
+    """
+
+    taulist = np.asarray(taulist)
+
+    dims = state0.dims[0]
+    A = _fix_op_type(A, dims)
+    B = _fix_op_type(B, dims)
+    C = _fix_op_type(C, dims)
```

Comment:
```suggestion
    A = qeye(dims) if A in [None, 1] else A
    B = qeye(dims) if B in [None, 1] else B
    C = qeye(dims) if C in [None, 1] else C
    A, B, C = QobjEvo(A), QobjEvo(B), QobjEvo(C)
```
I know this lead to some repetition in the code but I believe readability is much improved in this way. Besides, this is hopefully just one more line.